File: backend/internal/services/user_service.go
package services

import (
	"errors"
	"mini-meeting/internal/models"
	"mini-meeting/internal/repositories"
	"strings"

	"gorm.io/gorm"
)

type UserService struct {
	repo        *repositories.UserRepository
	meetingRepo *repositories.MeetingRepository
}

func NewUserService(repo *repositories.UserRepository, meetingRepo *repositories.MeetingRepository) *UserService {
	return &UserService{
		repo:        repo,
		meetingRepo: meetingRepo,
	}
}

func (s *UserService) GetUserByID(id uint) (*models.User, error) {
	user, err := s.repo.FindByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("user not found")
		}
		return nil, err
	}
	return user, nil
}

func (s *UserService) GetUserByEmail(email string) (*models.User, error) {
	email = strings.ToLower(email)
	user, err := s.repo.FindByEmail(email)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("user not found")
		}
		return nil, err
	}
	return user, nil
}

func (s *UserService) GetAllUsersPaginated(page, pageSize int, search string) (*models.PaginatedUsersResponse, error) {
	// Set default values
	if page < 1 {
		page = 1
	}
	if pageSize < 1 || pageSize > 100 {
		pageSize = 10
	}

	users, total, err := s.repo.FindAllPaginated(page, pageSize, search)
	if err != nil {
		return nil, err
	}

	// Calculate total pages
	totalPages := int(total) / pageSize
	if int(total)%pageSize > 0 {
		totalPages++
	}

	return &models.PaginatedUsersResponse{
		Data:       users,
		Total:      total,
		Page:       page,
		PageSize:   pageSize,
		TotalPages: totalPages,
	}, nil
}

func (s *UserService) UpdateUser(id uint, req *models.UpdateUserRequest) (*models.User, error) {
	user, err := s.repo.FindByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("user not found")
		}
		return nil, err
	}

	// Protect admin from being modified to regular user
	// Role cannot be changed through this endpoint
	if user.Role == "admin" && id == 1 {
		return nil, errors.New("cannot modify admin user")
	}

	// Update fields if provided
	if req.Name != "" {
		user.Name = req.Name
	}

	if err := s.repo.Update(user); err != nil {
		return nil, err
	}

	return user, nil
}

func (s *UserService) DeleteUser(id uint) error {
	// Check if user exists
	user, err := s.repo.FindByID(id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return errors.New("user not found")
		}
		return err
	}

	// Protect admin from deletion
	if user.Role == "admin" {
		return errors.New("cannot delete admin user")
	}

	// Delete all meetings created by this user
	if err := s.meetingRepo.DeleteByCreatorID(id); err != nil {
		return err
	}

	return s.repo.Delete(id)
}

func (s *UserService) FindOrCreateOAuthUser(provider models.Provider, providerID string, email string, name string, avatarURL string) (*models.User, error) {
	email = strings.ToLower(email)

	// First, try to find user by provider and provider ID
	user, err := s.repo.FindByProvider(provider, providerID)
	if err == nil && user != nil {
		// Update avatar URL if it changed
		if user.AvatarURL != avatarURL {
			user.AvatarURL = avatarURL
			s.repo.Update(user)
		}
		return user, nil
	}

	// Check if a user with this email already exists
	existingUser, err := s.repo.FindByEmail(email)
	if err == nil && existingUser != nil {
		// User exists with same email but different provider
		if existingUser.Provider != provider {
			// Already linked to another OAuth provider
			return nil, errors.New("account_exists_different_provider")
		}
	}

	// Create new OAuth user
	user = &models.User{
		Email:      email,
		Name:       name,
		Provider:   provider,
		ProviderID: providerID,
		AvatarURL:  avatarURL,
		Role:       "user",
	}

	if err := s.repo.Create(user); err != nil {
		return nil, err
	}

	return user, nil
}

func (s *UserService) LinkOAuthProvider(userID uint, provider models.Provider, providerID string, avatarURL string) error {
	user, err := s.repo.FindByID(userID)
	if err != nil {
		return errors.New("user not found")
	}

	// Check if already linked to this provider
	if user.Provider == provider && user.ProviderID == providerID {
		return nil // Already linked
	}

	// Check if provider is already in use by another account
	existingUser, err := s.repo.FindByProvider(provider, providerID)
	if err == nil && existingUser != nil && existingUser.ID != userID {
		return errors.New("provider already linked to another account")
	}

	// Update provider info
	return s.repo.UpdateProviderInfo(userID, provider, providerID, avatarURL)
}

func (s *UserService) GetUserByProvider(provider models.Provider, providerID string) (*models.User, error) {
	user, err := s.repo.FindByProvider(provider, providerID)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("user not found")
		}
		return nil, err
	}
	return user, nil
}

File: frontend/src/services/api/user.service.ts
import apiClient from "./client";
import type {
  User,
  SummarizerSession,
  UserResponse,
  PaginatedUsersResponse,
  UpdateUserRequest,
  UserUpdateResponse,
  CreateUserRequest,
  UserDeleteResponse,
} from "../../types/user.types";

export const userService = {
  /**
   * Get current authenticated user
   */
  getCurrentUser: async (): Promise<User> => {
    const response = await apiClient.get<UserResponse>("/users/me");
    return response.data.data;
  },

  /**
   * Update current user's profile
   */
  updateCurrentUser: async (data: UpdateUserRequest): Promise<User> => {
    const response = await apiClient.patch<UserUpdateResponse>("/users/me", data);
    return response.data.data;
  },

  /**
   * Get all users with pagination and search (Admin only)
   */
  getAllUsers: async (
    page: number = 1,
    pageSize: number = 10,
    search: string = ""
  ): Promise<PaginatedUsersResponse> => {
    const response = await apiClient.get<PaginatedUsersResponse>("/users", {
      params: { page, page_size: pageSize, search },
    });
    return response.data;
  },

  /**
   * Get user by ID (Admin only)
   */
  getUserById: async (id: number): Promise<User> => {
    const response = await apiClient.get<UserResponse>(`/users/${id}`);
    return response.data.data;
  },

  /**
   * Create new user (Admin only)
   */
  createUser: async (data: CreateUserRequest): Promise<User> => {
    const response = await apiClient.post<UserUpdateResponse>("/users", data);
    return response.data.data;
  },

  /**
   * Update user by ID (Admin only)
   */
  updateUser: async (id: number, data: UpdateUserRequest): Promise<User> => {
    const response = await apiClient.patch<UserUpdateResponse>(`/users/${id}`, data);
    return response.data.data;
  },

  /**
   * Get a specific summarizer session by ID for the current user
   */
  getSession: async (id: number): Promise<SummarizerSession> => {
    const user = await userService.getCurrentUser();
    const session = user.summarizer_sessions?.find((s) => s.id === id);
    if (!session) {
      throw new Error(`Session ${id} not found`);
    }
    return session;
  },

  /**
   * Delete user (Admin only)
   */
  deleteUser: async (id: number): Promise<void> => {
    await apiClient.delete<UserDeleteResponse>(`/users/${id}`);
  },
};
